xla/service/float_normalization.cc

```diff
@@ -564,6 +570,58 @@ Status FloatNormalizationVisitor::Preprocess(HloInstruction* hlo) {
   return OkStatus();
 }
 
+// We must avoid normalizing computations that have non-normalizing users
+// (e.g., all-reduce's computations should not be normalized). If a
+// computation is shared between normalizing and non-normalizing users, we will
+// clone the computation for the non-normalizing users so that it can be left
+// unchanged. This function clones the shared computations and returns the set
+// of non-normalizing computations that must be skipped by the visitor.
+absl::flat_hash_set<HloComputation*>
+CloneComputationsForNonNormalizingInstructions(
+    HloModule* module, std::vector<HloComputation*>& computations_to_visit) {
+  absl::flat_hash_set<HloComputation*> computations_to_skip;
+  // Let us first figure out what users each computation has.
+  struct ComputationUsers {
+    bool user_to_normalize_exists = false;
+    std::vector<HloInstruction*> users_to_skip_normalization;
+  };
+  absl::flat_hash_map<HloComputation*, ComputationUsers> computation_users;
+  for (auto& computation : computations_to_visit) {
+    for (auto* instruction : computation->instructions()) {
+      if (!instruction->has_called_computations()) continue;
```

Comment:
Nit: can we just remove this if-statement?

Reply:
Replaced with CallGraph now.